kip-0004.md

```diff
@@ -0,0 +1,104 @@
+```
+  KIP: 4
+  Layer: Consensus (hard fork), DAA
+  Title: Sparse Difficulty Windows
+  Author: Shai Wyborski <shai.wyborski@mail.huji.ac.il>
+          Michael Sutton <msutton@cs.huji.ac.il>
+          Georges Künzli <georges.kuenzli@dplanet.ch>
+  Status: implemented in testnet ([PR](https://github.com/kaspanet/rusty-kaspa/pull/186))
+```
+
+# Motivation
+The difficulty adjustment algorithm requires maintaining, for each block, a list of the N blocks in its past with the highest accumulated weight. When validating a block this list needs to be traversed to compute the various quantities required for computing the difficulty target. Since the length of the difficulty window corresponds to real-time considerations, the size of desired difficulty window is measured in seconds, not blocks. This means that if we increase the block rate by a factor of R, we increase the complexity of the difficulty adjustment by R^2 (the first factor is due to the fact that validating each block requires R times more steps, and the second factor is since blocks emerge R times faster).
+
+This is partially mitigated by the incremental nature of the difficulty window. Like the blue set, each block inherits its selected parent's difficulty window and then fixes it according to the blocks in its merge set. This converts some of the CPU overhead to RAM overhead, which would also increase by a factor of R.
+
+To reduce the overhead, the current Golang implementation uses window sizes that are not optimal (263 blocks for timestamp flexibility, 2641 blocks for DAA window). As a result, miners whose clock drifts into the future more than two minutes than the majority of the network have their block delayed decreasing the probability they will be blue. A secondary motivation for this proposal is to increase the timestamp flexibility (which requires increasing the DAA window accordingly) efficiently.
```

Comment:
maybe "Hence a secondary motivation..." ? reading the whole Motivation section is a bit confusing now, the ending paragraph makes it sound as if the sub-optimal sizes in current mainnet are the main motivation